proof/refine/ARM/Ipc_R.thy

```diff
@@ -4109,101 +4536,298 @@ lemma replyPush_sch_act_wf:
+  supply if_split [split del]
+  unfolding replyPush_def
+  apply wpsimp
+        apply (wpsimp wp: bindScReply_sym_refs_list_refs_of_replies'
+                          hoare_vcg_if_lift hoare_vcg_imp_lift' hoare_vcg_all_lift)
+       apply (rule_tac Q="(\<lambda>_ s.
+               sym_refs (list_refs_of_replies' s) \<and>
+               (\<forall>rptr scp. obj_at' (\<lambda>ko. scReply ko = Some rptr) scp s \<longrightarrow>
+                           obj_at' (\<lambda>ko. replyNext ko = Some (Head scp)) rptr s) \<and>
+               obj_at' (\<lambda>ko. replyPrev ko = None \<and> replyNext ko = None) replyPtr s)" in hoare_strengthen_post[rotated])
+        apply (fastforce split: if_splits simp del: comp_apply)[1]
+       apply (wpsimp wp: setReplyTCB_list_refs_of_replies' hoare_vcg_all_lift hoare_vcg_imp_lift'
+                         setReplyTCB_obj_at'_inv threadGet_wp)+
+  apply (intro conjI)
+     apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def)
+    apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def)
+   apply (clarsimp simp: sym_refs_asrt_def)
+   apply (rule ReplySchedContext_obj_at'_ref)
+    apply (fastforce simp: sym_refs_def dest: obj_at'_SCReply_ref)
+   apply (frule_tac p=xa in  obj_at_ko_at', clarsimp)
+   apply (frule (1) sc_ko_at_valid_objs_valid_sc')
+   apply (clarsimp simp: valid_sched_context'_def)
+  apply (erule (2) valid_replies'_no_tcb_not_linked)
+  done
+
+lemma bindScReply_if_live_then_nonz_cap':
+  "\<lbrace>if_live_then_nonz_cap'
+    and ex_nonz_cap_to' scPtr and ex_nonz_cap_to' replyPtr
+    and (\<lambda>s. \<forall>rp. obj_at' (\<lambda>ko. scReply ko = Some rp) scPtr s \<longrightarrow> obj_at' (\<lambda>ko. replyNext ko = Some (Head scPtr)) rp s )\<rbrace>
+   bindScReply scPtr replyPtr
+   \<lbrace>\<lambda>_. if_live_then_nonz_cap'\<rbrace>"
+  unfolding bindScReply_def
+  apply (simp (no_asm_use) add: getReplyTCB_def split del: if_split
+         | wp hoare_vcg_all_lift hoare_vcg_disj_lift hoare_vcg_imp_lift'
+              hoare_vcg_if_lift setReply_obj_at' setReplyTCB_obj_at'
+         | rule threadGet_wp)+
+  apply clarsimp
+  apply (erule if_live_then_nonz_capE')
+  apply (subgoal_tac "replyNext koc = Some (Head scPtr)")
+   apply (clarsimp simp: ko_wp_at'_def obj_at'_def projectKO_eq live_reply'_def projectKO_reply)
+  apply (clarsimp simp: ko_wp_at'_def obj_at'_def)
+  done
+
+lemma bindScReply_ex_nonz_cap_to'[wp]:
+  "bindScReply scPtr replyPtr \<lbrace>ex_nonz_cap_to' ptr\<rbrace>"
+  unfolding bindScReply_def
+        apply (simp (no_asm_use) add: getReplyTCB_def split del: if_split
+               | wp hoare_vcg_all_lift hoare_vcg_disj_lift hoare_vcg_imp_lift'
+                    hoare_vcg_if_lift setReply_obj_at' setReplyTCB_obj_at'
+               | rule threadGet_wp)+
+  apply clarsimp
   done
 
+crunches setThreadState, setReplyTCB
+  for obj_at'_sc[wp]: "\<lambda>s. Q (obj_at' (P :: sched_context \<Rightarrow> bool) scp s)"
+  (wp: crunch_wps)
```

Comment:
This seems to have been duplicated, I think you also have it earlier in this file.